### PR TITLE
i#2062: Support filtering of non-module blocks

### PR DESCRIPTION
**Pull request by abhinav92003**
Adds support for instrumenting non-module code instr-by-instr instead of whole-block-at-a-time. This is required for the L0 filter mode which instruments individual instructions rather than one pc entry for the whole block.

Under the new scheme, the modoffs field in each trace PC entry point to each non-mod instrs separately, rather than just the top of the non-mod block. Specifically, we store the cumulative encoding length of instrs stored to the encoding file prior to the recorded instr. This may potentially allow too few gencode blocks for JIT apps (we'd support 8G of JIT code); we could use multiple modidx to point to gencode (growing downward from PC_MODIDX_INVALID) to help with that in the future. Added a TODO.

Adds file type to the encoding file header. Adds a new encoding file version to indicate the presence of file type. Adds a new file type bit that denotes that the trace was filtered and that the module_mapper_t should interpret the modoffs as pointing to a single instr as described above. When this bit is not set, we use the existing scheme of interpreting the pc modoffs as just the non-mod block-idx.

Note that even under the new scheme, we still write the encoding file one mon-module block at a time because it is too inefficient to write one encoding_entry_t for just one non-module instr.

Modifies record_instr_encodings to skip writing anything to the encoding file for blocks without any app instr.

Adds a new variant of the tool.drcacheoff.gencode test that runs with the L0_filter enabled. Modifies the test to add a sequence of instrs that, without this fix, produces an error in raw2trace due to an apparently out-of-block memref.

Issue: #2062

### CODE REVIEW

File: clients/drcachesim/tracer/instru_offline.cpp [hunk 1]
```diff
@@ -480,6 +491,7 @@ offline_instru_t::record_instr_encodings(void *drcontext, app_pc tag_pc,
     log_(3, "%s: new block id " UINT64_FORMAT_STRING " for %p\n", __FUNCTION__,
          encoding_id_, tag_pc);
     per_block->id = encoding_id_++;
+    per_block->encoding_length_start = encoding_length_;
```
Comment (derekbruening) [inline clients/drcachesim/tracer/instru_offline.cpp:494]: It took me a while (offline discussion) to understand why we didn't simply store one instruction per encoding entry and have the encoding ID be per-instruction: it seems much simpler at first, avoiding all this encoding length tracking and avoiding the raw2trace split code paths and new raw2trace map.  The only downside seems to be the extra space from the 3 64-bit fields per encoding entry which is a lot -- though for small-gencode apps like those just using VDSO it still seems better that way as the encoding entry count is so small.

Please update the PR description to explain *why* this adds this extra complexity: to save space.

Please add comment(s) in the code too: maybe in trace_entry.h and also in raw2trace.h.

**Review by derekbruening — APPROVED** [review]

File: api/docs/release.dox [hunk 2]
```diff
@@ -142,6 +142,9 @@ changes:
    refers to timestamps and direct switches, which is what most users should want.
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply api/docs/release.dox:142]: Done.

File: clients/drcachesim/common/trace_entry.h [hunk 3]
```diff
@@ -969,7 +975,28 @@ typedef union {
...
+    /**
+     * Default encoding file type.
+     */
+    ENCODING_FILE_TYPE_DEFAULT = 0x0,
+    /**
+     * This encoding file type tells the module_mapper_t that the non-module PC
+     * entries in the trace correspond to an individual instr. The modoffs field is
+     * interpreted as the cumulative encoding length of all instrs written to the
+     * encoding file before the recorded instr. If this file type is not set, then
+     * non-module PC entries are interpreted as the non-mod block's idx.
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/common/trace_entry.h:996]: Done.

File: clients/drcachesim/tests/burst_gencode.cpp [hunk 4]
```diff
@@ -287,38 +304,40 @@ post_process()
...
+        "-satisfy_w_xor_x "
 #endif
-                   "-stderr_mask 0xc -client_lib ';;-offline"))
+        "-stderr_mask 0xc -client_lib ';;-offline";
+    env += add_env;
+    if (!my_setenv("DYNAMORIO_OPTIONS", env.c_str()))
         std::cerr << "failed to set env var!\n";
     code_generator_t gen(false);
-    std::cerr << "pre-DR init\n";
+    std::cerr << "pre-DR init\n" << std::flush;
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/burst_gencode.cpp:324]: Even on my x86 Linux workstation I saw some cases where the output was just missing.  I see the following output every few runs:

```
pre-DR init
pre-DR start
preall done
```

Adding the std::flush solved it.

When I replaced it with a std::endl, now I'm seeing weird duplication:
```
pre-DR init
pre-DR start
pre-DR detachall done
all done
```

File: clients/drcachesim/tracer/raw2trace.cpp [hunk 5]
```diff
@@ -349,17 +349,29 @@ module_mapper_t::do_encoding_parsing()
...
+    uint64_t encoding_file_version = *reinterpret_cast<uint64_t *>(map_at);
+    map_at += sizeof(uint64_t);
+    if (encoding_file_version > ENCODING_FILE_VERSION)
         return "Encoding file has invalid version";
-    size_t offs = sizeof(uint64_t);
-    while (offs < file_size) {
-        encoding_entry_t *entry = reinterpret_cast<encoding_entry_t *>(map_start + offs);
-        if (entry->length < sizeof(encoding_entry_t))
+    if (encoding_file_version >= ENCODING_FILE_VERSION_HAS_FILE_TYPE) {
+        uint64_t encoding_file_type = *reinterpret_cast<uint64_t *>(map_at);
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tracer/raw2trace.cpp:359]: Done.

File: clients/drcachesim/tests/burst_gencode.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/tests/burst_gencode.cpp:324]: I believe that if you are seeing a difference with that flush call, your C++ implementation violates the standard.  std::cerr has unitbuf set (https://en.cppreference.com/w/cpp/io/manip/unitbuf) which flushes after each operation.  Adding std::flush should be a nop.  And std::endl is supposed to be the same as a newline plus a call to flush (https://en.cppreference.com/w/cpp/io/manip/endl).

File: clients/drcachesim/tracer/tracer.cpp [hunk 6]
```diff
@@ -2319,10 +2319,10 @@ drmemtrace_client_main(client_id_t id, int argc, const char *argv[])
...
         DR_ASSERT(MAX_INSTRU_SIZE >= sizeof(offline_instru_t));
         placement = dr_global_alloc(MAX_INSTRU_SIZE);
-        instru = new (placement)
-            offline_instru_t(insert_load_buf_ptr, &scratch_reserve_vec,
-                             file_ops_func.write_file, module_file, encoding_file,
-                             op_disable_optimizations.get_value(), instru_notify);
+        instru = new (placement) offline_instru_t(
+            insert_load_buf_ptr, &scratch_reserve_vec, file_ops_func.write_file,
+            module_file, encoding_file, op_disable_optimizations.get_value(),
+            op_L0D_filter.get_value() || op_L0I_filter.get_value(), instru_notify);
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tracer/tracer.cpp:2325]: I suppose we can just use the new scheme for the whole trace. Added `|| op_L0_filter_until_instrs`.

I'm seeing a crash while handling the ud2 (the test makes one on purpose). Debugging.

[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tracer/tracer.cpp:2325]: Created #6474 for this. This seems to happen in burst_gencode with L0_filter_until_instrs even without this fix.

File: clients/drcachesim/tracer/instru_offline.cpp (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/tracer/instru_offline.cpp:494]: Added it to the PR description and to trace_entry.h.

File: clients/drcachesim/tracer/tracer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tracer/tracer.cpp:2327]: I tried debugging i#6474 but couldn't root-cause it completed in a reasonable amount of time. I added TODOs for now.

File: clients/drcachesim/tests/burst_gencode.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/burst_gencode.cpp:324]: For some reason it was harder to reproduce this. The test is no longer failing without the std::flush.

When it was failing, I ran under strace and saw the following:

```
write(2, "pre-DR detach\0all done", 22pre-DR detachall done) = 22
..
write(2, "\n", 1
)                       = 1
write(2, "all done", 8all done)                 = 8
write(2, "\n", 1
)                       = 1
```

[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/burst_gencode.cpp]: No actually it's still there (found some inconsistent behaviour when running the test manually outside Cmake). I filed https://github.com/DynamoRIO/dynamorio/issues/6490 with a reliable way to reproduce the issue. and I'm using std::flush in this PR with a TODO.

**Comment by abhinav92003** [comment]
x86-32 failures are #6417.

**Pull request merged**